cylc/flow/network/__init__.py

```diff
@@ -30,34 +35,71 @@
     CylcError,
     CylcVersionError,
     ServiceFileError,
-    WorkflowStopped
+    WorkflowStopped,
 )
 from cylc.flow.hostuserutil import get_fqdn_by_host
 from cylc.flow.workflow_files import (
     ContactFileFields,
-    KeyType,
-    KeyOwner,
     KeyInfo,
+    KeyOwner,
+    KeyType,
+    get_workflow_srv_dir,
     load_contact_file,
-    get_workflow_srv_dir
 )
 
+
+if TYPE_CHECKING:
+    # BACK COMPAT: typing_extensions.TypedDict
+    # FROM: Python 3.7
+    # TO: Python 3.11
+    from typing_extensions import TypedDict
+
+
 API = 5  # cylc API version
 MSG_TIMEOUT = "TIMEOUT"
 
+if TYPE_CHECKING:
+    class ResponseDict(TypedDict, total=False):
+        """Structure of server response messages.
 
-def encode_(message):
-    """Convert the structure holding a message field from JSON to a string."""
-    try:
-        return json.dumps(message)
-    except TypeError as exc:
-        return json.dumps({'errors': [{'message': str(exc)}]})
+        Confusingly, has similar format to GraphQL execution result.
+        But if we change this now we could break compatibility for
+        issuing commands to/receiving responses from workflows running in
+        different versions of Cylc 8.
+        """
+        data: object
+        """For most Cylc commands that issue GQL mutations, the data field will
+        look like:
+        data: {
+        <mutationName1>: {
+            result: [
+            {
+                id: <workflow/task ID>,
+                response: [<success_bool>, <message>]
+            },
+            ...
+            ]
+        }
+        }
+        but this is not 100% consistent unfortunately
+        """
+        error: Union[Exception, str, dict]
+        """If an error occurred that could not be handled.
+        (usually a dict {message: str, traceback?: str}).
+        """
+        user: str
+        cylc_version: str
+        """Server (i.e. running workflow) Cylc version.
+
+        Going forward, we include this so we can more easily handle any future
+        back-compat issues."""
 
 
-def decode_(message):
-    """Convert an encoded message string to JSON with an added 'user' field."""
+def load_server_response(message: str) -> 'ResponseDict':
+    """Convert a JSON message string to dict with an added 'user' field."""
     msg = json.loads(message)
-    msg['user'] = getpass.getuser()  # assume this is the user
+    if 'user' not in msg:
```

Comment:
Under what circumstances would this be set already, and should we trust it if it is?

Reply:
This `user` field seems to be used by the [`WorkflowRuntimeServer.receiver()`](https://github.com/cylc/cylc-flow/blob/21d18ba0b22c3e729345aa30e539ee3cee1211c7/cylc/flow/network/server.py#L289) method. It is set by the [`WorkflowReplier.listener()`](https://github.com/cylc/cylc-flow/blob/21d18ba0b22c3e729345aa30e539ee3cee1211c7/cylc/flow/network/replier.py#L103) method. In the server main loop it is listening for zmq messages and then assuming that they are from the current user.

https://github.com/cylc/cylc-flow/blob/21d18ba0b22c3e729345aa30e539ee3cee1211c7/cylc/flow/network/replier.py#L100-L103

It says "authenticating the user in the process" but this is always the current user, so I don't know why we would always set the current user in this dict.

Might be one for @dwsutherland?

Reply:
> Might be one for @dwsutherland?

Nah, that one I put in.

The user field should be filled in for requests inbound from the UIS (if Cylc Hub is being used) so we can log the user who requested the action.

If the request didn't come via the UIS, then we log it as originated from the workflow owner (i.e. the person who actioned this command has control over the owner's user account and can act on its behalf).

Reply:
Actually from debugging I'm pretty sure this is vestiges of authorisation functionality that was removed/changed: https://github.com/cylc/cylc-flow/pull/3845/files#diff-97d852e5ae5195e580b347c04b37bd759726680ab8e19dfaa079288d05280b45R46. The `user` field is not actually used anymore.

What is used is `auth_user`: https://github.com/cylc/cylc-flow/blob/21d18ba0b22c3e729345aa30e539ee3cee1211c7/cylc/flow/network/resolvers.py#L717-L730

Reply:
> What is used is auth_user

Ach, ok.

I think what's happened here is:
* The user field as added at the ZMQ layer to support multi-user interaction (following on the pattern of Cylc 7).
* Cylc 8 implemented authorisation at the UIS layer, the `auth_user` field was added to GraphQL.
* The user field could still have use at the CLI layer.
* However, we decided that a multi-user CLI would go through the UIS anyway (i.e. via GraphQL) so there should be no present or future use case for having this field at the ZMQ level.

Reply:
I've force-pushed to leave the `msg['user']` the way it was, and taken out the removal of the unused auth stuff, to be put on an 8.5.0 PR